util: Fix compilation errors in support/lockedpool.cpp

jkczyz opened the pull request:
Changes in #12048 cause a compilation error in Arena::walk() when
ARENA_DEBUG is defined. Specifically, Arena's chunks_free map was
changed to have a different value type.

Additionally, missing includes cause other compilation errors when
ARENA_DEBUG is defined.

Reproduced with:

make CPPFLAGS=-DARENA_DEBUG

jkczyz committed:
Fix compilation errors in support/lockedpool.cpp

Changes in #12048 cause a compilation error in Arena::walk() when
ARENA_DEBUG is defined. Specifically, Arena's chunks_free map was
changed to have a different value type.

Additionally, missing includes cause other compilation errors when
ARENA_DEBUG is defined.

Reproduced with:

make CPPFLAGS=-DARENA_DEBUG

Empact commented:
How about a test case?

jkczyz commented:
@Empact Would it be sufficient to remove `#ifdef ARENA_DEBUG` from this code to ensure the compilation error is caught in the future? Not sure if this goes against conventions for debug code.

Empact commented:
No the guards are there for a reason, we just need to exercise the guarded code at some point.

May not be practical, but one possibility is treating one of the Travis test runs with ARENA_DEBUG.

jkczyz committed:
Define ARENA_DEBUG in Travis test runs

The definition and uses of Arena::walk() are compiled only if
ARENA_DEBUG is defined. Configure Travis to define ARENA_DEBUG so
compilation errors do not go unnoticed.

jkczyz commented:
I defined ARENA_DEBUG for a couple environments. Wasn't sure which environments, so I just added it to ones that were already setting CPPFLAGS. The Travis CI build failures seem unrelated.

I had to update the Arena's test to use allocated memory to prevent a segfault. Figured this may have been needed as I observed it locally.

laanwj commented on src/test/allocator_tests.cpp:
```diff
@@ -15,11 +15,9 @@ BOOST_FIXTURE_TEST_SUITE(allocator_tests, BasicTestingSetup)
     const size_t synth_size = 1024*1024;
-    Arena b(synth_base, synth_size, 16);
+    std::unique_ptr<char[]> synth_base(new char[synth_size]());
+    Arena b(synth_base.get(), synth_size, 16);
```
Using a fake address here was on purpose, it would catch accesses to the underlying memory by causing a segfault.

laanwj replied:
I'm slightly confused. I don't think `walk` is supposed to access the contents of the memory, it is supposed to be a debugging function that prints the allocation metadata, addresses and such, which is intentionally stored outside of the arena (to save on scarce page-locked memory). If it accesses, or even dereferences, the underlying memory that seems like a bug to me.

jkczyz replied:
I'm not sure why, but Inserting `base` into `std::cout` inside `printchunk` causes the segfault when executed with `make CPPFLAGS=-DARENA_DEBUG check`:

```console
Entering test module "Bitcoin Core Test Suite"
test/allocator_tests.cpp(14): Entering test suite "allocator_tests"
test/allocator_tests.cpp(16): Entering test case "arena_tests"
0xunknown location(0): fatal error: in "allocator_tests/arena_tests": memory access violation at address: 0x080ffc10: no mapping at fault address                                        
test/allocator_tests.cpp(16): last checkpoint: "arena_tests" entry.
Test is aborted
test/allocator_tests.cpp(16): Leaving test case "arena_tests"; testing time: 8196us
Test is aborted
test/allocator_tests.cpp(14): Leaving test suite "allocator_tests"; testing time: 8231us
Test is aborted
```

The segfault does not occur if I remove `base` from being inserted into the stream.

jkczyz replied:
Turns out this segfault can be avoided instead by changing `printchunk` to take `void*` rather than `char*`, which is what I had been doing in #16195. Updated in 30fb598.

laanwj commented:
Thanks for catching this and fixing the build with `-DARENA_DEBUG`, not 100% sure about the build system and test and other changes.

fanquake commented:
Concept ACK. @jkczyz could you rebase this?

Empact commented:
Concept ACK

jkczyz commented:
> Concept ACK. @jkczyz could you rebase this?

Done.

jkczyz committed:
Fix segfault in allocator_tests/arena_tests

The test uses reinterpret_cast<void*> on unallocated memory. Using this
memory in printchunk as char* causes a segfault, so have printchunk take
void* instead.

laanwj commented:
ACK 30fb598737f6efb7802d707a1fa989872e7f8b7b

I like how minimal this change has become.

fanquake reviewed (approved):
ACK 30fb598737f6efb7802d707a1fa989872e7f8b7b - thanks for following up jkczyz.

```bash
src/test/test_bitcoin --run_test=allocator_tests
Running 3 test cases...
0x00000000x80ffc10 0x00000000000003f0 0x1

0x00000000x8000000 0x00000000000ffc10 0x0


0x00000000x8000000 0x0000000000100000 0x0

0x00000000x80ffc80 0x0000000000000200 0x1
0x00000000x80ffe80 0x0000000000000100 0x1
0x00000000x80fff80 0x0000000000000080 0x1

0x00000000x8000000 0x00000000000ffc80 0x0

0x00000000x80ffc80 0x0000000000000200 0x1
0x00000000x80ffe80 0x0000000000000100 0x1

0x00000000x80fff80 0x0000000000000080 0x0
0x00000000x8000000 0x00000000000ffc80 0x0

0x00000000x80fff80 0x0000000000000080 0x1
0x00000000x80ffc80 0x0000000000000200 0x1

0x00000000x80ffe80 0x0000000000000100 0x0
0x00000000x8000000 0x00000000000ffc80 0x0


0x00000000x8000000 0x0000000000100000 0x0
```
